Client start option

MetalStormGhost opened the pull request:
Added a new settings option: Default start type for Player aircraft. Refactored to use the new StartType enum from game.ato.starttype. Resolves #1567 but also allows other start types for Player flights besides hot start.

MetalStormGhost committed:
Added a missing call to commit_cargo_ship_losses when committing mission results, fixing a problem which caused lost ground units on sinked cargo ships to not register.

MetalStormGhost committed:
Merge branch 'dcs-liberation:develop' into develop

MetalStormGhost committed:
Merge branch 'dcs-liberation:develop' into develop

MetalStormGhost committed:
Merge branch 'dcs-liberation:develop' into develop

MetalStormGhost committed:
Added a new option: Default start type for Player aircraft

MetalStormGhost committed:
Reorganized the signals and got rid of the circular dependency. QFlightCreator now also behaves correctly when selecting/deselecting player pilots. Resolves #1567

MetalStormGhost committed:
Reverted my changes to settings.py and QSettingsWindow.py in preparation for the settings rework.

MetalStormGhost committed:
Merge branch 'dcs-liberation:develop' into develop

MetalStormGhost committed:
Merge branch 'dcs-liberation:develop' into develop

MetalStormGhost committed:
Updated Marianas campaigns to DCS World 2.7.7.14727 open beta.

MetalStormGhost committed:
Fetch upstream from dcs-liberation:develop Oct 23, 2021
Merge branch 'develop' of https://github.com/MetalStormGhost/dcs_liberation into client-start-options

MetalStormGhost committed:
Option to select default start type for Player aircraft refactored to use the new StartType enum from game.ato.starttype. Resolves #1567

MetalStormGhost committed:
Restored missing Settings import in packagebuilder.py

MetalStormGhost committed:
Added campaign designers a possibility to configure the number of parking slots at airfields by (optionally) defining "override_aircraft_parking" in the campaign yaml (for each airfield separately). This will override the actual number of parking slots in DCS, so doing so will make all aircraft at the airfield start on runway.

This is intended to increase the operational feasibility of certain airfields in campaigns. For example, Jiroft in the Persian Gulf (Iran interior) only has three parking slots.

MetalStormGhost committed:
Revert "Added campaign designers a possibility to configure the number of parking slots at airfields by (optionally) defining "override_aircraft_parking" in the campaign yaml (for each airfield separately). This will override the actual number of parking slots in DCS, so doing so will make all aircraft at the airfield start on runway."

This reverts commit e7d7c293cb916ecf108105400e10e269d908b79c.

MetalStormGhost committed:
Revert "Updated Marianas campaigns to DCS World 2.7.7.14727 open beta."

This reverts commit ce10bdea11c2b0d0e589d092157245e0b1dd5ea7.

MetalStormGhost committed:
Merge branch 'client-start-options' of https://github.com/MetalStormGhost/dcs_liberation into client-start-options

DanAlbert commented on game/commander/packagebuilder.py:
```diff
@@ -61,6 +64,12 @@ def plan_flight(self, plan: ProposedFlight) -> bool:
+        # If this is a client flight, set the start_type again to match the configured default
+        # https://github.com/dcs-liberation/dcs_liberation/issues/1567
+        if flight.roster is not None and flight.roster.player_count > 0:
+            flight.start_type = self.settings.default_start_type_client
```
Probably safer to add a separate `player_start_type` field to the `Flight` constructor so it can deal with this internally rather than needing to remember to do this in every location? Does mean that `Flight` needs to also get the `Settings` object but that's probably going to be true sooner or later anyway.

MetalStormGhost replied:
I tried to do that but I ran into a very nasty looking circular import problem...

MetalStormGhost replied:
@DanAlbert Could I ask for your help with getting Settings object access in Flight?

DanAlbert commented on qt_ui/windows/mission/flight/QFlightCreator.py:
```diff
@@ -178,7 +182,7 @@ def create_flight(self) -> None:
             roster.max_size,
             task,
-            self.start_type.currentText(),
+            StartType(self.start_type.currentText()),
```
Instead of doing this, you can add values to the combo box:

```
for start_type in StartType:
    self.start_type.addItem(start_type.value, start_type)
```

And then use `self.start_type.currentData()` instead.

DanAlbert commented on qt_ui/windows/mission/flight/QFlightCreator.py:
```diff
@@ -237,3 +241,20 @@ def update_max_size(self, available: int) -> None:
+
+        for i, st in enumerate([b for b in ["Cold", "Warm", "Runway", "In Flight"]]):
+            if start_type.value == st:
+                self.start_type.setCurrentIndex(i)
```
Don't do this IMO. If they player has manually set this field we shouldn't be resetting it just because they added or removed a pilot (especially because adding more slots might add pilots unintentionally).

Could add yet another "default start type for manually created flights" as well if that's workload reducing, but I think the main value of the client default is in making the auto ATO more auto.

MetalStormGhost replied:
How about I add a flag if the user has manually set this field? I could then use the flag to prevent resetting their changes.

DanAlbert replied:
Maybe it would be better to replace `Flight.start_type: StartType` with `Flight.custom_start_type: Optional[StartType]`, and the UI can instead show "Default (Cold)" or whatever, and the default type can be resolved as-needed? The advantage would be that the UI code and behavior is better, plus changing the settings can have an effect _this turn_ instead of next turn. The downside is that means that any time someone alters that setting every package TOT needs to be re-evaluated (the ASAP ones of course, but also all the others because switching from hot to cold might make early takeoffs suddenly be negative).

The trouble with the flag in the UI is that it's only in the UI so if you close and open the window it no longer knows if the start type was an explicit choice or not. Storing that UI specific data in the save game is... gross.

I think I still like my original idea best for the time being. It's the simplest and I suspect it's more than enough to solve the problem.

MetalStormGhost replied:
I could try your approach of showing "Default (Cold)" etc, especially now since we have that new StartType class. I assume iterating over all flights and updating their TOT is trivial, but I haven't tried it out yet.

DanAlbert replied:
I should probably mention that some things I have planned for 6 or 7 will make this obsolete anyway, so the simple answer is going to be the right answer.

DanAlbert commented on qt_ui/windows/mission/flight/settings/QFlightStartType.py:
```diff
@@ -39,7 +49,26 @@ def __init__(self, package_model: PackageModel, flight: Flight):
+            if self.flight.start_type.value == st:
+                self.start_type.setCurrentIndex(i)
+
+        self.package_model.update_tot()
```
Same.

MetalStormGhost committed:
Merge branch 'dcs-liberation:develop' into develop

MetalStormGhost committed:
Moved default_start_type_client from packagebuilder.py to flight.py as suggested.

Merge branch 'develop' of https://github.com/MetalStormGhost/dcs_liberation into client-start-options

MetalStormGhost committed:
Revert "Moved default_start_type_client from packagebuilder.py to flight.py as suggested."

This reverts commit 86b4994d714dd5caa92806cdc14cd2bd8a2a70e9, reversing
changes made to 3f957d4d1b8af9526582ccd9825a1ed171820b35.

MetalStormGhost committed:
Merge branch 'dcs-liberation:develop' into develop

MetalStormGhost committed:
Merge branch 'dcs-liberation:develop' into client-start-options

MetalStormGhost committed:
Converted Flight.start_type: StartType to Flight.custom_start_type: Optional[StartType]. Still very much a draft and has some issues with handling the StartType enum and None values. Also, Flight needs access to Settings but that will require fixing a circular dependency.

MetalStormGhost committed:
Merge branch 'dcs-liberation:develop' into develop

MetalStormGhost committed:
Merge branch 'develop' of https://github.com/MetalStormGhost/dcs_liberation into client-start-options

MetalStormGhost committed:
Now sets the "Default (Cold)" text in the start type combobox. Still needs access to settings in flight.py to work properly.

MetalStormGhost committed:
flight.start_type -> flight.get_start_type

MetalStormGhost committed:
Revert "Merge branch 'develop' of https://github.com/MetalStormGhost/dcs_liberation into client-start-options"

This reverts commit 19c30d708a8d374c907f2995eaad0d87040bba89, reversing
changes made to 1e6f1ecbde7faa57e4c4e880d29da2736d702245.

MetalStormGhost committed:
Fixed game\ato\flight.py:131: error: Incompatible return value type (got "str", expected "StartType")

RndName commented:
Is this PR still maintained?

MetalStormGhost commented:
> Is this PR still maintained?

I ran into a circular import problem when trying to access the new setting for client start option in the Flight object. I stopped working on it at that time, but I could clean this up and complete this if I could get that circular import figured out. If this was C++, I'd have already done this months ago.

Having spent the time learning how the dependencies work with Liberation, I could take another look at this.

DanAlbert commented:
I don't think there's any point in doing so. Like I said, this is going to be obsolete soon anyway because start type won't exist once turns are gone
